### ISSUE TITLE
[new rule] prefer toHaveBeenCalledTimes()

### ISSUE THREAD

**Issue by robcresswell** [jest-community/eslint-plugin-jest#109]
This is a new rule suggestion; it would be nice to enforce `toHaveBeenCalledTimes()` over `toHaveBeenCalled()`, to ensure functions are only being executed a set amount of times. It's unlikely making this fixable would be useful, as we would have to make an assumption about the number of times, which could well be incorrect.

I'm happy to make a PR to support this, but don't have a definite timeframe. I thought it was worth opening the issue for discussion first.

**Comment by SimenB** [comment]
I think that's a great idea!

It would also be cool to fix thing like `expect(someMock.mock.calls.length).toBe(3)` into `expect(someMock).toHaveBeenCalledTimes(3)`, but it might give some false positives?

**Comment by robcresswell** [comment]
That's a good idea, I hadn't thought of that. I can't think of any examples of how that would cause a false positive off the top of my head, but thats only off a couple thousand lines of experience so hardly a big sample. If nobody beats me to it, I'll have a go at writing the rule out. Should get time tomorrow evening.

**Comment by SimenB** [comment]
e.g.:

```js
const someMock = {
  mock: {
    calls: ['something', 'weird']
  }
}
```

Not sure how realistic that is, though.

**Comment by edbrannin** [comment]
@SimenB I don't quite follow your last comment.  Were you trying to illustrate a situation where changing `expect(foo.mock.calls.length).toBe(x)` to `expect(foo).toHaveBeenCalledTimes(x)` might break something?

(Because I came here wondering if someone had done that rule yet)

**Comment by SimenB** [comment]
`toHaveBeenCalledTimes` verifies it's called with a mock function - so if you have an object that's _not_ a mock function, autofixing would break your test

**Comment by G-Rath** [comment]
@SimenB sounds like this is a fix we should *suggest* 😉

**Comment by SimenB** [comment]
hah, you _called_ it

**Comment by benmonro** [comment]
I was just going to suggest this feature as well.  the rule should also handle `expect(some.mock.calls).toHaveLength(x)` and `expect(some.mock.calls[0][0]).toBe("foo")`

first one could be fixable, 2nd... might be trickier.
